observeOn: allow configurable buffer size

srvaroa opened the pull request:
The observeOn operator is backed by a small queue of 128 slots that may
overflow quickly on slow producers.  This could only be avoided by
adding a backpressure operator before the observeOn (not only
inconvenient, but also taking a perf. hit as it forces hops between two
queues).

This patch allows modifying the default queue size on the observeOn
operator.

Fixes: #3751
Signed-off-by: Galo Navarro <anglorvaroa@gmail.com>

akarnokd commented on src/main/java/rx/Observable.java:
```diff
@@ -6308,13 +6309,42 @@ public final void forEach(final Action1<? super T> onNext, final Action1<Throwab
+     * @see #observeOn(Scheduler)
+     * @see #observeOn(Scheduler, boolean)
+     */
+    public final Observable<T> observeOn(Scheduler scheduler, int bufferSize) {
```
Please add `observeOn(Scheduler, boolean delayError, int bufferSize)` overload as well.

akarnokd replied:
I still miss the overload `observeOn(Scheduler, boolean delayError, int bufferSize)`

akarnokd commented on src/test/java/rx/internal/operators/OperatorObserveOnTest.java:
```diff
@@ -523,14 +523,20 @@ public boolean hasNext() {
         final CountDownLatch latch = new CountDownLatch(1);
+        // randomize buffer size, note that underlaying implementations may be tuning the real size to a power of 2
+        // which can lead to unexpected results when adding excess capacity (e.g.: see ConcurrentCircularArrayQueue)
+        final int capacity = (int)Math.pow(2, new Random().nextInt(10));
```
There is no need for randomized capacity. If you want to verify multiple sizes, loop over a power-of-2 set.

artem-zinnatullin replied:
oh yes, please don't use `Random` in tests, they will be non-reproducible

akarnokd commented:
:+1:

artem-zinnatullin commented on src/test/java/rx/internal/operators/OperatorObserveOnTest.java:
```diff
@@ -523,14 +523,21 @@ public boolean hasNext() {
     }
 
+    /**
+     * {@link BackpressureTests.testObserveOnWithSlowConsumer} also covers the case with default buffer size.
```
nit: this link won't work, I guess

artem-zinnatullin commented:
:+1: though i'd leave old test for default buffer size and just add a new one

srvaroa commented:
Rebased, fixed the javadoc link, and a spelling mistake.

@artem-zinnatullin I reused this one to avoid redundancy as the code path is virtually the same, and there are tests (the ones referenced in the @link) which use the default size. Let me know if you still prefer the 2 cases and I'll fix that.

artem-zinnatullin commented:
Yeah, I understand, just afraid that one may change linked  tests and we won't cover overload. 

// one day we'll setup code coverage, I hope

akarnokd commented on src/test/java/rx/internal/operators/OperatorObserveOnTest.java:
```diff
@@ -523,14 +523,21 @@ public boolean hasNext() {
         final CountDownLatch latch = new CountDownLatch(1);
+        // randomize buffer size, note that underlying implementations may be tuning the real size to a power of 2
+        // which can lead to unexpected results when adding excess capacity (e.g.: see ConcurrentCircularArrayQueue)
+        final int capacity = (int)Math.pow(2, new Random().nextInt(10));
```
I agree with @artem-zinnatullin, leave this test as it was and create a new one with a loop over some queue sizes:

```java
for (int i = 1; i <= 1024; i = i * 2) {
   final int capacity = i;
   // ...
}
```

srvaroa commented:
Agh thanks @akarnokd, I didn't notice I changed computers so I was back on the 1st patch and lost the overload. Resubmitted based on the one with all overloads, and added the extra test instead of rewriting it. Hopefully fine now.

akarnokd commented:
:+1:

artem-zinnatullin commented on src/test/java/rx/internal/operators/OperatorObserveOnTest.java:
```diff
@@ -581,6 +581,69 @@ public void onNext(Integer t) {
     }
 
     @Test
+    public void testQueueFullEmitsErrorWithVaryingBufferSize() {
```
nit: `test` prefix is not needed

srvaroa commented:
I've seen some failures on the test locally and there is also a bug in the test (the loop should start at 2, not 1). I can't look into it right now but will take a look asap.

stevegury commented:
:+1:

akarnokd commented:
I'm merging this. The changes are algorithmically solid so I'm not sure what failures you see locally.

srvaroa commented:
Thanks @akarnokd. The failure was no error notifications emitted at random sizes, but I did notice Travis didn't complain. Ping me if it surfaces.
